src/transform/src/cse/anf.rs

```diff
@@ -121,6 +124,45 @@ impl Bindings {
     ) -> Result<(), crate::TransformError> {
         self.checked_recur_mut(|this| {
             match relation {
+                MirRelationExpr::LetRec {
+                    ids,
+                    values,
+                    body,
+                    max_iters: _,
+                } => {
+                    // Extend this.rebindings with (id, id) pairs for each
+                    // recursive ID before descending into values and body.
+                    this.rebindings.extend(ids.iter().map(|id| {
+                        let new_id = this.id_offset + u64::cast_from(this.bindings.len());
```

Comment:
Because in the `Let` case we now that as a consequence of the bottom-up traversal the `$value` for `cte $x = $value` is already in the `Get $y` form, so we can extend the context with `rebind($x, $y)`.

In the `LetRec` case in general every `value` from the associated `values` can reference every binding from the associated `ids`, so it is not so simple. We need to allocate an `id_offset` range for all `ids` upfront and extend the `rebindings` map before we descend into the first `value`.

Reply:
It seems that at rendering time we are importing / exporting declared indexes

https://github.com/MaterializeInc/materialize/blob/f2145eea8e36107f332b3fbddf96b4c84f03cfa3/src/compute/src/render/mod.rs#L289-L315

so maybe it is possible to also to somehow call `context.export_index_iterative(...)` with the head of each `object: dataflow.objects_to_build`?

Reply:
Aha, ok, thanks. So it's because there might be a tower of Lets there due to the `populate_expression` call for the value.